streamparse/ext/invoke.py

```diff
@@ -193,96 +208,138 @@ def submit_topology(name=None, env_name="prod", par=2, options=None,
-
-    # pre-submit hooks for invoke and fabric
-    user_invoke, user_fabric = get_user_tasks()
-    pre_submit_invoke = getattr(user_invoke, "pre_submit", None)
-    if callable(pre_submit_invoke):
-        pre_submit_invoke(name, env_name, env_config)
-    pre_submit_fabric = getattr(user_fabric, "pre_submit", None)
-    if callable(pre_submit_fabric):
-        pre_submit_fabric(name, env_name, env_config)
+    # Check if we need to maintain virtualenv during the process
+    use_venv = env_config.get('use_virtualenv', True)
+    if use_venv:
+        activate_env(env_name)
 
-    config["virtualenv_specs"] = config["virtualenv_specs"].rstrip("/")
+    _pre_submit_hooks(name, env_name, env_config)
 
-    create_or_update_virtualenvs(
-        name, "{}/{}.txt".format(config["virtualenv_specs"], name)
-    )
-    python_path = '/'.join([env_config["virtualenv_root"],
-                           name, "bin", "python"])
+    if use_venv:
+        config["virtualenv_specs"] = config["virtualenv_specs"].rstrip("/")
+        create_or_update_virtualenvs(
+            name, "{}/{}.txt".format(config["virtualenv_specs"], name)
+        )
 
     # Prepare a JAR that doesn't have Storm dependencies packaged
     topology_jar = jar_for_deploy()
 
     print('Deploying "{}" topology...'.format(name))
-    with ssh_tunnel(env_config["user"], host, 6627, port):
-        print("ssh tunnel to Nimbus {}:{} established."
-              .format(host, port))
-
-        if force and not is_safe_to_submit(name):
-            print("Killing current \"{}\" topology.".format(name))
-            _kill_topology(name, run_kwargs={"hide": "both"}, wait=wait)
-            while not is_safe_to_submit(name):
-                print("Waiting for topology {} to quit...".format(name))
-                time.sleep(0.5)
-
-            print("Killed.")
-
-        jvm_opts = [
-            "-Dstorm.jar={}".format(topology_jar),
-            "-Dstorm.options=",
-            "-Dstorm.conf.file=",
-        ]
-        os.environ["JVM_OPTS"] = " ".join(jvm_opts)
-        cmd = ["lein",
-               "run -m streamparse.commands.submit_topology/-main",
-               topology_file]
-        if debug:
-            cmd.append("--debug")
-        cmd.append("--option 'topology.workers={}'".format(par))
-        cmd.append("--option 'topology.acker.executors={}'".format(par))
+    # Use ssh tunnel with Nimbus or use host/port for Thrift connection
+    if check_if_use_ssh_for_nimbus(env_config):
+        with ssh_tunnel(env_config["user"], host, 6627, port):
+            print("ssh tunnel to Nimbus {}:{} established."
+                  .format(host, port))
+            _kill_existing_topology(name, force, wait)
+            _submit_topology(name, topology_file, topology_jar,
+                             env_config, par, options, debug)
+            _post_submit_hooks(name, env_name, env_config)
+            return
+
+    # This part doesn't use SSH tunnel at all
```

Comment:
Minor preference thing, but could you just put this part in an `else:` block and get rid of the `return` in the `if` block?